script/copy_deduplicate

```diff
@@ -358,47 +373,72 @@ def list_live_tables(client, pool, project_id, except_tables, only_tables):
+        with self.client() as client:
+            func(client, *args)
+
+
 def main():
     args = parser.parse_args()
 
-    client = bigquery.Client()
+    # create a queue for balancing load across projects
+    client_q = ClientQueue(args.billing_projects, args.parallelism)
 
     with ThreadPool(args.parallelism) as pool:
-        live_tables = list_live_tables(
-            client=client,
-            pool=pool,
-            project_id=args.project_id,
-            except_tables=args.except_tables,
-            only_tables=args.only_tables,
-        )
-
-        job_args = [
-            args
-            for jobs in pool.starmap(
-                get_query_job_configs,
-                [
-                    (
-                        client,
-                        live_table,
-                        args.date,
-                        args.dry_run,
-                        args.slices,
-                        args.priority,
-                        args.preceding_days,
-                    )
-                    for live_table in live_tables
-                ],
+        with client_q.client() as client:
+            live_tables = list_live_tables(
+                client=client,
+                pool=pool,
+                project_id=args.project_id,
+                except_tables=args.except_tables,
+                only_tables=args.only_tables,
             )
-            for args in jobs
-        ]
 
-        # preserve job_args order so results stay sorted by stable_table for groupby
-        results = pool.starmap(run_deduplication_query, job_args, chunksize=1)
-        copy_args = [
-            (client, stable_table, [query_job for _, query_job in group])
+            query_jobs = [
+                (run_deduplication_query, *args)
+                for jobs in pool.starmap(
+                    get_query_job_configs,
+                    [
+                        (
+                            client,  # only use one client to create temp tables
+                            live_table,
+                            date,
+                            args.dry_run,
+                            args.slices,
+                            args.priority,
+                            args.preceding_days,
+                        )
+                        for live_table in live_tables
+                        for date in args.dates
+                    ],
+                )
+                for args in jobs
+            ]
+
+        # preserve query_jobs order so results stay sorted by stable_table for groupby
+        results = pool.starmap(client_q.with_client, query_jobs, chunksize=1)
+        copy_jobs = [
+            (copy_join_parts, stable_table, [query_job for _, query_job in group])
             for stable_table, group in groupby(results, key=lambda result: result[0])
         ]
-        pool.starmap(copy_join_parts, copy_args)
+        pool.starmap(client_q.with_client, copy_jobs, chunksize=1)
```

Comment:
I just tried running this on one of the main_v4 backfill tables, and it looks like it ran all 100 slices across 8 projects as I had configured, but then failed with

```
	Traceback (most recent call last):
  File "./heka-backfill/py-multi-copydedupe", line 447, in <module>
    main()
  File "./heka-backfill/py-multi-copydedupe", line 441, in main
    for stable_table, group in groupby(results, key=lambda result: result[0])
  File "./heka-backfill/py-multi-copydedupe", line 440, in <listcomp>
    (copy_join_parts, stable_table, [query_job for _, query_job in group])
  File "./heka-backfill/py-multi-copydedupe", line 441, in <lambda>
    for stable_table, group in groupby(results, key=lambda result: result[0])
TypeError: 'NoneType' object is not subscriptable
```

So perhaps the return value of the query jobs has changed.

Reply:
fixed